Manage GitHub resources with Terraform

bilbof opened the pull request:
This adds a new GitHub deployment which enables us to manage GitHub resources with Terraform.

Currently it simply permits alphagov repositories with the tag 'govuk' to push images to ECR, by providing access to credentials.

There are various additional features we can add to simplify and automate the configuration of new repositories and deploying new applications to the Kubernetes cluster. This should tie in nicely with the ECR module, though that module could also use the github provider to fetch the repositories to create ECR registries, which would mean we don't need to manually add to a list of applications in the module when we add a new app.

One downside is that folks that use this module need to have org-level permissions. Ideally this deployment would be run by an automated process that sits in-cluster and watches GitHub for tag changes.

sengi commented on terraform/deployments/github/README.md:
````diff
@@ -0,0 +1,23 @@
+  ```shell
+  GITHUB_TOKEN=<token> gds aws govuk-production-poweruser -- terraform plan
+  ```
+4. Plan Terraform
````
```suggestion
4. Apply Terraform
```

sengi commented on terraform/deployments/github/README.md:
```diff
@@ -0,0 +1,23 @@
+# GOV.UK GitHub
```
I think this needs a more specific name and a description of what specifically it's for, especially as we already have [another config-as-code tool that's configuring GitHub](https://github.com/alphagov/govuk-saas-config/tree/master/github). We'll also need to signpost the two tools from each other - and from devdocs - otherwise it's gonna be painful for anyone trying to figure out how to change GitHub settings in future.

sengi commented on terraform/deployments/github/main.tf:
```diff
@@ -0,0 +1,41 @@
+#
+
+data "github_repositories" "govuk" {
+  # TODO: Add more granular tags to repositories (i.e. `dockerised`, `replatformed`)
```
It's not clear what this TODO is intended to accomplish - is it really worth requiring people to set another tag on (almost) every GOV.UK repo (including new ones and the time taken troubleshooting it when someone forgets, etc.)?

sengi reviewed (approved):
I think this is a good solution, though it introduces a potentially confusing piece of tech debt: splitting our GitHub config across two very different places. I think if we signpost it nicely (here, any relevant places in govuk-saas-config and wherever saas-config is mentioned in devdocs) then it'll be fine though.

sengi commented:
FWIW I think this is a cleaner approach than govuk-saas-config - TF is a better fit than some Ruby scripts for this kind of thing.

bilbof commented:
Thanks @sengi. I've updated the commit, created a tech debt card, and added [a PR to flag this on govuk-saas-config](https://github.com/alphagov/govuk-saas-config/pull/101).
